dataset/src/main/scala/frameless/TypedEncoder.scala

```diff
@@ -296,30 +342,119 @@ object TypedEncoder {
     }
   }
 
+  implicit def arrayEncoder[A : ClassTag](
+    implicit
+    underlying: TypedEncoder[A]
+  ): TypedEncoder[Array[A]] = new CollectionEncoder[Array, A]() {
+    val nullable: Boolean = false
+
+    val sourceDataType: DataType = FramelessInternals.objectTypeFor[Array[A]](classTag)
+
+    val targetDataType: DataType = DataTypes.createArrayType(underlying.targetDataType)
+
+    def constructorFor(path: Expression): Expression = arrayData(path)
+
+    def extractorFor(path: Expression): Expression = {
+      // if source `path` is already native for Spark, no need to `map`
+      if (ScalaReflection.isNativeType(underlying.sourceDataType)) {
+        NewInstance(
+          classOf[GenericArrayData],
+          path :: Nil,
+          dataType = ArrayType(underlying.targetDataType, underlying.nullable)
+        )
+      } else {
+        MapObjects(underlying.extractorFor, path, underlying.sourceDataType)
+      }
+    }
+  }
+
+  implicit def mapEncoder[A : ClassTag, B : ClassTag](
+    implicit
+    encodeA: TypedEncoder[A],
+    encodeB: TypedEncoder[B]
+  ): TypedEncoder[Map[A, B]] = new TypedEncoder[Map[A, B]] {
+    val nullable: Boolean = false
+    val sourceDataType = FramelessInternals.objectTypeFor[Map[A, B]]
+    val targetDataType = MapType(encodeA.targetDataType, encodeB.targetDataType, encodeB.nullable)
+
+    // needed to construct Map using `ArrayBasedMapData.toScalaMap` (avoids an extra MapObjects)
+    private val arrayA = arrayEncoder[A]
+    private val arrayB = arrayEncoder[B]
+
+    // needed to extract keys/values from map - toArray requires an implicit and can't be called from codegen Java
+    private val vectorA = vectorEncoder[A]
+    private val vectorB = vectorEncoder[B]
+
+
+    private def wrap(arrayData: Expression) = {
+      StaticInvoke(
+        scala.collection.mutable.WrappedArray.getClass,
+        FramelessInternals.objectTypeFor[Seq[_]],
+        "make",
+        arrayData :: Nil)
+    }
+
+    def constructorFor(path: Expression): Expression = {
+      val keyArrayType = ArrayType(encodeA.targetDataType, false)
+      val keyData = wrap(arrayA.constructorFor(Invoke(path, "keyArray", keyArrayType)))
+
+      val valueArrayType = ArrayType(encodeB.targetDataType, encodeB.nullable)
+      val valueData = wrap(arrayB.constructorFor(Invoke(path, "valueArray", valueArrayType)))
+
+      StaticInvoke(
+        ArrayBasedMapData.getClass,
+        sourceDataType,
+        "toScalaMap",
+        keyData :: valueData :: Nil)
+    }
+
+    def extractorFor(path: Expression): Expression = {
+      val keys =
+        Invoke(
```

Comment:
Given generated code in on critical path, I think it completely makes sense avoid allocating `Vector` with custom functions `Iterator[Int] => Array[Int]` for primitives and `Iterator[A] => Array[A]` for boxed types.